### PR TITLE
Reducing boilerplate code for Redux

### PR DESCRIPTION
**Pull request by hugohaggmark**
The goal with the POC is to show a way to reduce the amount of boilerplate code used to create a strongly typed redux solution with actions, action creators, reducers and tests.

``+`` Much less boilerplate code
``-`` Non Redux standard api

### actionCreatorFactory
Used to create an action creator with the following signature 
```typescript
{ type: string , (payload: T): {type: string; payload: T;} }
```
where the ``type`` string will be ensured to be unique and ``T`` is the type supplied to the factory.

#### Example
```typescript
export const someAction = actionCreatorFactory<string>('SOME_ACTION').create();

// later when dispatched
someAction('this rocks!');
```
```typescript
// best practices, always use an interface as type
interface SomeAction { data: string; } 
export const someAction = actionCreatorFactory<SomeAction>('SOME_ACTION').create();

// later when dispatched
someAction({ data: 'best practices' });
```
```typescript
// declaring an action creator with a type string that has already been defined will throw
export const someAction = actionCreatorFactory<string>('SOME_ACTION').create();
export const theAction = actionCreatorFactory<string>('SOME_ACTION').create(); // will throw
```

### noPayloadActionCreatorFactory
Used when you don't need to supply a payload for your action. Will create an action creator with the following signature 
```typescript
{ type: string , (): {type: string; payload: undefined;} }
```
where the ``type`` string will be ensured to be unique.

#### Example
```typescript
export const noPayloadAction = noPayloadActionCreatorFactory('NO_PAYLOAD').create();

// later when dispatched
noPayloadAction();
```
```typescript
// declaring an action creator with a type string that has already been defined will throw
export const noPayloadAction = noPayloadActionCreatorFactory('NO_PAYLOAD').create();
export const noAction = noPayloadActionCreatorFactory('NO_PAYLOAD').create(); // will throw
```

### reducerFactory
Fluent API used to create a reducer. (same as implementing the standard switch statement in Redux)
(there is a similar fluent api to ``reducerFactory`` in the code called ``reducerFactoryEx`` just with another syntax, we should pick the best one 😄if we decide to use this pattern)

#### Example
```typescript
interface ExampleReducerState { data: string[]; }

const intialState: ExampleReducerState = { data: [] };

export const someAction = actionCreatorFactory<string>('SOME_ACTION').create();
export const otherAction = actionCreatorFactory<string[]>('Other_ACTION').create();

export const exampleReducer = reducerFactory<ExampleReducerState>(intialState)
// addMapper is the function that ties an action creator to a state change
.addMapper({ 
    // action creator to filter out which mapper to use
    filter: someAction, 
    // mapper function where the state change occurs
    mapper: (state, action) => ({ ...state, data: state.data.concat(action.payload) }),
  })
// a developer can just chain addMapper functions until reducer is done
.addMapper({
    filter: otherAction,
    mapper: (state, action) => ({ ...state, data: action.payload }),
  })
  .create(); // this will return the reducer
```

#### Typing limitations
There is a challenge left with the mapper function that I can not solve with TypeScript. The signature of a mapper is

```typescript
<State, Payload>(state: State, action: ActionOf<Payload>) => State
```
If you would to return an object that is not of the state type like the following mapper  
```typescript
mapper: (state, action) => ({ nonExistingProperty: ''}),
```
Then you would receive the following compile error
```shell
[ts] Property 'data' is missing in type '{ nonExistingProperty: string; }' but required in type 'ExampleReducerState'. [2741]
```
But if you return an object that is spreading state and add a non existing property type like the following mapper  
```typescript
mapper: (state, action) => ({ ...state, nonExistingProperty: ''}),
```
Then you would not receive any compile error.

If you want to make sure that never happens you can just supply the State type to the mapper callback like the following mapper:
```typescript
mapper: (state, action): ExampleReducerState => ({ ...state, nonExistingProperty: 'kalle' }),
```
Then you would receive the following compile error
```shell
[ts]
Type '{ nonExistingProperty: string; data: string[]; }' is not assignable to type 'ExampleReducerState'.
  Object literal may only specify known properties, and 'nonExistingProperty' does not exist in type 'ExampleReducerState'. [2322]
```

### reducerTester
Fluent API that simplifies the testing of reducers

#### Example
```typescript
reducerTester()
        .givenReducer(someReducer, initialState)
        .whenActionIsDispatched(someAction('reducer tests'))
        .thenStateShouldEqual({ ...initialState, data: 'reducer tests' });
```

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
This is a really big improvement, great stuff! My online gripe is the move from a native switch statement to an array find call, would be nice if there was a way to use an index lookup there instead.

File: public/app/core/redux/reducerFactory.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      throw new Error(`There is already a Mappers defined with the type ${config.filter.type}`);
+    }
+
+    allMapperConfigs.push(config);
+
+    return instance;
+  };
+
+  const create = (): Reducer<State, ActionOf<any>> => (state: State = initialState, action: ActionOf<any>): State => {
+    const mapperConfig = allMapperConfigs.filter(config => config.filter.type === action.type)[0];
```
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/core/redux/reducerFactory.ts]: Awesome suggestion, I'll look into index lookup tomorrow

**Comment by dprokop** [comment]
This is a great improvement to our workflow!

File: public/app/core/redux/reducerFactory.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/core/redux/reducerFactory.ts]: Pushed a new version now

**Pull request merged**